std/functional.d

```diff
@@ -1400,6 +1393,16 @@ unittest
     assert(fact(10) == 3628800);
 }
 
+// Issue 20099
+@system unittest // not @safe due to memoize
+{
+    int i = 3;
+    alias a = memoize!((n) => i + n);
```

Comment:
I think having to add the param types for lambdas is acceptable to work around this: `memoize!((int n) => i + n)`, and that the danger illustrated by n8sh above isn't worth the trouble.

Reply:
For user-friendliness, we could catch that case though, e.g.:
```D
static if (fun.stringof.length > 8 && fun.stringof[0 .. 8] == "__lambda" && !isCallable!fun)
    static assert(false, "memoized lambdas require explicit parameter types");
```
=>
```
onlineapp.d(6): Error: static assert:  "memoized lambdas require explicit parameter types"
onlineapp.d(29):        instantiated from here: `memoize!((n) => i + n)`
```